### PR TITLE
web: Generate catalog of all media content for caching in PWA

### PR DESCRIPTION
**Pull request by Vadim-Karpenko**
The first step on working on [https://github.com/renpy/renpy-build/issues/53](https://github.com/renpy/renpy-build/issues/53).
This code creates a minified `pwa_catalog.json` file in the destination build that has the following structure:
```
{"files": { "<file_path>": "<hash>", ... }, "version": "your-game-<timestamp>"}
```
`files` contains all paths in the `game` folder, and will allow us to download all files at once, for offline use.
The `version` will let us track if the catalog is updated to the current version of the service worker. If not, we'll create a warning in the console, to let the dev know that it needs to be updated

### CODE REVIEW

**Comment by qinlili23333** [comment]
I think file hash should be contained in catalog, this can help for delta update. Maybe MD5 is a good choice, but for files larger than 1M maybe only calculate the first 1M part MD5.

**Comment by Vadim-Karpenko** [comment]
> I think file hash should be contained in catalog, this can help for delta update. Maybe MD5 is a good choice, but for files larger than 1M maybe only calculate the first 1M part MD5.
> 
> And in service worker update process, we can do like this: Step 1: fetch latest catalog Step 2: if exist cached old catalog then compare file's hash and get a list of files need to update, or if it's a new installation all necessary files will add to the list Step 3: download files in the list Or we can put the update process in the first time the game is launched with new service worker, since if it take too long time in the install event of service worker it may get fail. Just like most mobile games, silent update for main component(apk for Android, and service worker for PWA), and trigger resource update in next launch.

Yeah, makes sense. It will decrease the amount of HEAD requests needed to keep everything up-to-date. And I don’t think that media files will be changed after the build is complete (manually by the user), like it can happen to core files (e.g index.html). In most cases it will be rebuilt and hashes will be updated too. 
I’ll prepare the code.

**Comment by qinlili23333** [comment]
>  And I don’t think that media files will be changed after the build is complete (manually by the user), like it can happen to core files (e.g index.html). In most cases it will be rebuilt and hashes will be updated too. 

Maybe we can also provide a standalone tool for user to generate catalog manually after file change. Put catalog build code as a seperate file "build-catalog.py", invoke this file in building and user can also manually invoke this file after file change.

**Review by qinlili23333 — APPROVED** [review]

File: launcher/game/web.rpy [hunk 1]
```diff
@@ -338,8 +361,9 @@ init python:
 
         # Use re to slugify the game name, avoiding use of 3rd party libraries
         slugified_name = re.sub(r'\W+', '-', p.dump['build']['display_name']).lower()
+        version = "{}-{}".format(slugified_name, int(time.time()))
         # Replace the default cache name with the game name + current timestamp
-        service_worker = service_worker.replace('renpy-web-game', "{}-{}".format(slugified_name, int(time.time())))
+        service_worker = service_worker.replace('renpy-web-game', version)
```
Comment (qinlili23333) [inline launcher/game/web.rpy:407]: I would like to seperate app name and version in service worker, and seperately replace these two values in build.
in service worker:
```
var APP_PREFIX = '||GAME_NAME||'
var VERSION = '||GAME_VERSION||'
var CACHE_NAME = APP_PREFIX + VERSION;
```
And in build:
```
service_worker = service_worker.replace('||GAME_NAME||', slugified_name)
service_worker = service_worker.replace('||GAME_VERSION||', version)
```
And we need a way to get version of service worker in main webpage. I think we can do like this.
```
if (event.request.url.indexOf("getVersionWorker") > 0) {
    event.respondWith(new Response(VERSION));
    return;
}
```
So we can get version just like this.
```
let currentVersion = await (await fetch("getVersionWorker")).text();
```

**Review by qinlili23333 — CHANGES_REQUESTED** [review]

**Comment by renpytom** [comment]
Having tried the PWA out, I think the caching here is way too aggressive. 

Right now, if I launch a game in the Ren'Py launcher, change the game, rebuild it, and reload the web page, it will get the old content on the first pass through. I have to then hit reload to get access to the latest version of the game. That's unacceptable for development, and probably unacceptable for production - we don't want to get into a scenario where a creator puts out a fix, and then there are still bugs in it. 

My thinking is that the PWA should probably fetch a catalog from the server, first thing it does. If can use this catalog to determine which files can be pulled from the cache, and which can't. If fetching the catalog fails, run entirely out of the cache - otherwise, prefer the server unless we can confirm the cache is up to date.

**Comment by qinlili23333** [comment]
> My thinking is that the PWA should probably fetch a catalog from the server, first thing it does. If can use this catalog to determine which files can be pulled from the cache, and which can't. If fetching the catalog fails, run entirely out of the cache - otherwise, prefer the server unless we can confirm the cache is up to date.

Maybe we can setup a `version.json`, containing latest version and acceptable versions, just like this structure.
```
{
"latest":"1.3.2",
"acceptable":["1.3.0","1.3.1"]
}
```
And if version is latest or an network error occurs, just use cache directly. If version is in acceptable, use cache and do update in background if network allows (these non-essential updates should not take on mobile networks especially when data saver is on, we can detect by `navigator.connection`, and background update should only take effect in next time launch). If version is too old, just do force update before launch game. This checking action should put in webpage instead of service worker, and we should only check once in a period (maybe in 2 hours ?).
Defaultly, in build we only set `latest` to current version and set an empty array in `acceptable`, but developers can also manually set acceptable versions.

This is also why I want to add a `ForceNoCache` query detect in service worker. Requests with `ForceNoCache` key should always fetch from origin server without check cache. e.g. A fetch of `./config.json` will first detect cache, but a fetch of `./config.json?ForceNoCache=1` will directly request server.

File: launcher/game/web.rpy (hunk 1 above)
Comment (Vadim-Karpenko) [reply launcher/game/web.rpy:427]: The version is just a timestamp of when the game was built, not the version of the game, so not sure if we need to separate them

**Comment by Vadim-Karpenko** [comment]
> Having tried the PWA out, I think the caching here is way too aggressive.
> 
> Right now, if I launch a game in the Ren'Py launcher, change the game, rebuild it, and reload the web page, it will get the old content on the first pass through. I have to then hit reload to get access to the latest version of the game. That's unacceptable for development, and probably unacceptable for production - we don't want to get into a scenario where a creator puts out a fix, and then there are still bugs in it.
> 
> My thinking is that the PWA should probably fetch a catalog from the server, first thing it does. If can use this catalog to determine which files can be pulled from the cache, and which can't. If fetching the catalog fails, run entirely out of the cache - otherwise, prefer the server unless we can confirm the cache is up to date.

Spent the entire morning working on this, looks like it was the desired behavior from PWA devs. Tried a lot of things, but ended up reloading a page automatically if an update is found. Not the best solution, so I will look into it more in the coming days

**Pull request merged**